File: subscriber_member_function.cpp
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
using std::placeholders::_1;

// 定义一个MinimalSubscriber类，继承自rclcpp::Node
class MinimalSubscriber : public rclcpp::Node
{
  public:
    // MinimalSubscriber的构造函数
    MinimalSubscriber()
    : Node("minimal_subscriber")
    {
      // 创建一个订阅者，订阅名为"topic"的主题，队列大小为10，并绑定topic_callback函数
      subscription_ = this->create_subscription<std_msgs::msg::String>(
      "topic", 10, std::bind(&MinimalSubscriber::topic_callback, this, _1));
    }

  private:
    // topic_callback函数，用于处理接收到的消息
    void topic_callback(const std_msgs::msg::String::SharedPtr msg) const
    {
      // 使用RCLCPP_INFO宏和节点日志记录器输出接收到的消息内容
      RCLCPP_INFO(this->get_logger(), "I heard: '%s'", msg->data.c_str());
    }
    // 定义一个订阅者指针
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
};

// main函数，程序的入口点
int main(int argc, char * argv[])
{
  // 初始化ROS 2
  rclcpp::init(argc, argv);
  // 创建MinimalSubscriber类的共享指针，并使用rclcpp::spin函数保持节点运行
  rclcpp::spin(std::make_shared<MinimalSubscriber>());
  // 关闭ROS 2
  rclcpp::shutdown();
  // 返回0，表示程序成功执行
  return 0;
}
//

File: subscriber_member_function.py
import rclpy
from rclpy.node import Node

from std_msgs.msg import String

# 定义一个名为MinimalSubscriber的类，继承自Node
class MinimalSubscriber(Node):

    def __init__(self):
        # 调用父类Node的构造函数，并设置节点名称为'minimal_subscriber'
        super().__init__('minimal_subscriber')
        
        # 创建一个订阅者，订阅的话题类型为String，话题名称为'topic'，队列大小为10
        # 并设置回调函数为self.listener_callback
        self.subscription = self.create_subscription(
            String,
            'topic',
            self.listener_callback,
            10)
        
        # 防止订阅者变量未使用而引发警告
        self.subscription

    # 定义订阅者回调函数
    def listener_callback(self, msg):
        # 在日志中打印接收到的消息内容
        self.get_logger().info('I heard: "%s"' % msg.data)


# 主函数
def main(args=None):
    # 初始化rclpy，这是ROS2客户端库的Python接口
    rclpy.init(args=args)

    # 创建MinimalSubscriber类的实例
    minimal_subscriber = MinimalSubscriber()

    # 进入事件循环，等待回调函数被调用
    rclpy.spin(minimal_subscriber)

    # 显式销毁节点，这是一个可选步骤，因为当节点对象被垃圾回收时，节点会自动销毁
    minimal_subscriber.destroy_node()
    
    # 关闭rclpy，清理资源
    rclpy.shutdown()


# 当脚本作为主程序运行时，调用main函数
if __name__ == '__main__':
    main()
